base/channels.jl

```diff
@@ -457,3 +457,22 @@ function iterate(c::Channel, state=nothing)
 end
 
 IteratorSize(::Type{<:Channel}) = SizeUnknown()
+
+"""
+    Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` Tasks spawned by `Threads.@spawn`. This function
+will wait for all internally spawned tasks to complete before returning.
+"""
+function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
+    tasks = map(1:ntasks) do _
+        Threads.@spawn begin
+            for item in channel
+                wait(Threads.@spawn f(item))
+            end
+        end
+    end
+    Base.sync_end(tasks)
+    return nothing
+end
```

Comment:
Two suggestions:

* Use `@sync` instead of `sync_end`. It's recently changed to accept `Channel`. So current PR doesn't work anymore. I think it's better to use public interface even internally.
* Add a simple implementation of stop-all-on-error behavior.

```suggestion
function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads())
    stop = Threads.Atomic{Bool}(false)
    @sync for _ in 1:ntasks
        Threads.@spawn try
            for item in channel
                _waitspawn(f, item)
                stop[] && break  # do this _after_ f(item) to avoid loosing `item`
            end
        catch
            stop[] = true
            rethrow()
        end
    end
    return nothing
end

# Avoid registering the task in outer @sync:
_waitspawn(f, x) = wait(Threads.@spawn f(x))
```

Reply:
Thanks!